### PR TITLE
ALTER RESOURCE GROUP SET CONCURRENCY N

### PR DESCRIPTION
**Pull request by xiong-gang**
Increase the 'concurrency' limit can take effect immediately, and the
queueing transactions can be woken up. Decrease the 'concurrency' is
different, if the new limit is smaller than the number of current running
transactions, ALTER statement won't cancel the running transactions to
the limit. Therefore, we use column 'proposed' in pg_resgroupcapability
to represent the effective limit, and use column 'value' to record the
historical limit.
For example, we have a resource group with concurrency=3, and there
are 3 running transactions and 3 queueing transactions. If we alter the
concurrency to 2, 'proposed' will be updated to 2 and 'value' will stay
as 3. When one running transaction is finished, it won't wake up the
transactions in the queue as the current concurrency is 2. If we execute
the statement again to alter the concurrency to 2, it will update the
'value' column to 2, and the 'value' is consistent with 'proposed'
again.

Signed-off-by: Richard Guo <riguo@pivotal.io>

### CODE REVIEW

**Comment by xiong-gang** [comment]
@amilkh please take a look at this pr.

**Comment by amilkh** [comment]
@xiong-gang thanks for tagging me.

Glad to see that you are doing TDD! A few points:
1. The failing test should be written before writing code. I noticed the tests were added after the code was implemented.
2. I suggest you break up the .out files to be as small as possible. `src/test/isolation2/expected/resource_group.out` can be split into 2 or 3 files, and put into a `src/test/isolation2/expected/rg` folder.
3. Could you add some tests for edge cases, such as `ALTER RESOURCE GROUP rg_concurrency_test SET CONCURRENCY -1;` or `ALTER RESOURCE GROUP rg_concurrency_test SET CONCURRENCY 'string value';` or `ALTER RESOURCE GROUP rg_concurrency_test SET CONCURRENCY 9999999999999999999999999999999999999999999999999;`?
4. Is the isolation2 tester the best place to put these tests? Not sure, but seems like this is legacy and ICW is better?
5. Make sure to consider how these new tests can run in parallel so our CI doesn't slow down. Adding a new box to Concourse called "resource group tests" might be good if you plan to add time-consuming RG tests.
6. The tests you are adding now are functional tests. Writing unit tests on your code functions would be the next step, but it might be quite challenging to get that infrastructure set up on your end!

File: src/backend/commands/resgroupcmds.c [hunk 1]
```diff
@@ -800,19 +979,21 @@ insertTupleIntoResCapability(Relation rel,
 /*
  * Retrive capability value from pg_resgroupcapability
  */
-static text
-getCapabilityForGroup(int groupId, int type)
+static void
+getResgroupCapabilityEntry(int groupId, int type, char **value, char **proposed)
```
Comment (pivotal-ning-yu) [inline src/backend/commands/resgroupcmds.c]: Should the caller hold the lock(s) before calling this function?

File: src/backend/parser/gram.y [hunk 2]
```diff
@@ -1037,6 +1038,7 @@ stmt :
 			| AlterTypeStmt
 			| AlterUserSetStmt
 			| AlterUserStmt
+			| AlterResourceGroupStmt
```
Comment (pivotal-ning-yu) [inline src/backend/parser/gram.y]: Should we keep this list sorted?

File: src/backend/tcop/utility.c [hunk 3]
```diff
@@ -1747,6 +1748,9 @@ ProcessUtility(Node *parsetree,
 		case T_DropResourceGroupStmt:
 			DropResourceGroup((DropResourceGroupStmt *) parsetree);
 			break;
+		case T_AlterResourceGroupStmt:
```
Comment (pivotal-ning-yu) [inline src/backend/tcop/utility.c]: I noticed that most commands have the `case` lines in the order of CREATE, ALTER and DROP, shall we do the respect the same order?

File: src/backend/commands/resgroupcmds.c [hunk 4]
```diff
@@ -315,13 +316,98 @@ DropResourceGroup(DropResourceGroupStmt *stmt)
...
+				(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+				 errmsg("must be superuser to alter resource groups")));
+
+	if (stmt->concurrency < RESGROUP_CONCURRENCY_UNLIMITED)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_LIMIT_VALUE),
+				 errmsg("concurrency limit cannot be less than %d", RESGROUP_CONCURRENCY_UNLIMITED)));
+
+	/* any access to pg_resgroupcapability should be blocked during ALTER */
+	LockRelationOid(ResGroupCapabilityRelationId, AccessExclusiveLock);
```
Comment (pengzhout) [inline src/backend/commands/resgroupcmds.c]: Is there a potential dead lock here:
Assume there are two transactions A and B, A is altering resource group , B is creating resource group.

T1:  A get AccessExclusiveLock on ResGroupCapabilityRelationId
T2: B get ExclusiveLock on ResGroupRelationId and try to acquire RowExclusiveLock on ResGroupCapabilityRelationId within updateResgroupCapability().
T3: A try to acquire RowExclusiveLock on RowExclusiveLock and then deadlock.

Comment (amilkh) [inline src/backend/commands/resgroupcmds.c]: Would suggest you add an isolation test for this and fix it if it gets stuck.

**Comment by xiong-gang** [comment]
@amilkh Sorry for the late reply.
Isolation test is a test framework for concurrent scenario, currently we put the concurrent test cases in isolation2/resource_group and put the non-concurrent test cases in icw/resource_group.
As the resource group is shared relation across databases, we lean to not run it in parallel for now.
I am planning to split the tests in smaller ones and put in one folders, as another PR https://github.com/greenplum-db/gpdb/pull/2346 already did this, i will merge it later.
Will add some edge tests in ICW.
Thanks for the review and the suggestions about TDD:)

File: src/backend/commands/resgroupcmds.c (hunk 4 above)
Comment (xiong-gang) [inline src/backend/commands/resgroupcmds.c]: Good point. Fixed.

**Comment by amilkh** [comment]
@xiong-gang no problem. If you want tips on TDD, feel free to tag one of us again!

**Pull request merged**